andela/jubilee-backend
# #167733643 Email verification functionality

#### What does this PR do?
Add email verification functionality to the App, ensuring a user gets a verification email after a successful registration.

#### Description of Task to be completed?
- Install SendGrid, a third-party package for sending in-app emails.
- Implement methods for signing and verifying authentication token in `helpers.js` (a utility class file).
- Implement a method for generating verification links.
- Implement a method for sending verification emails in the `mailer.js` file.
- Add the email verification method to the signup implementation.
- Refactor the routing setup of the current codebase
- Create a route and an endpoint `api/auth/verify` for validating an email verification link.
- Implement a method in the `UserService class` for updating the account status of a user as verified, in the event of a successful verification through the aforementioned route.
- Write unit tests for the functionalities mentioned above.


#### How should this be manually tested?
- Clone repo and change into the working directory
- Switch to the ft-email-verification-167733643 branch and run `npm install`
- Create a .env file and specify the variables described in the .env.sample inside this file.
- Run `npm run prestart` to create tables in your database.
- Once the database tables are created, run `npm run start:dev` to start the development server.
- Launch Postman and make a post required to the signup route `http://localhost:3000/api/auth/signup` using the field described in the swagger document of the project.
- Upon successful signup, visit your registration mail and check for an email from the app, following the instruction described in the mail.
- Upon successful verification, a JSON response should be received with updated user parameters.
- To run the unit tests, run `npm test`.
- Document feature using swagger 

#### What are the relevant pivotal tracker stories?
#167733643

#### Screenshot

<img width="783" alt="test-passing" src="https://user-images.githubusercontent.com/40744698/63655109-97036e00-c77b-11e9-9926-94bc53fe849b.PNG">
<img width="946" alt="email" src="https://user-images.githubusercontent.com/40744698/63655113-a8e51100-c77b-11e9-8884-a271220c6140.PNG">
